src/rollback_to_stable/rts_btree.c

```diff
@@ -494,7 +518,8 @@ __rts_btree_ondisk_fixup_key(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip,
          * rollback to stable operation.
          */
         F_SET(upd, WT_UPDATE_RESTORED_FROM_HS);
```

Comment:
I think the dryrun mode should not be changing the `upd` structure. It feels like every place where this code alters the behavior of a statistic without changing an actual cleanup probably deserves careful review.

Reply:
I don't entirely agree with this: this `upd` is local to this function until we insert it into the update chain below via `__rts_btree_row/col_modify`, and those functions wrap the insert in a dryrun check.

I agree on that stats though, and I've tried to be careful there. The specific bit of code you're commenting on looks that way since we update the stats earlier when we have a bit more context about the operation, e.g. `__rts_btree_row/col_modify` doesn't know whether we're restoring a tombstone to history store. So we sometimes have the stats updates a long way from where the update is performed.

Reply:
Got it - I wonder if instead of having these stat increment calls wrapped in an `if` statement you could define a macro that hides the difference, something like:

```
#define WT_STAT_RTS_DATA_INCR(session, stat)     \
    do {                                                                                                                    \
        if (!S2C(session)->rts->dryrun)                                                                  \
             WT_STAT_CONN_DATA_INCR(session, stat);                                       \
    while (0)
```
Then the call becomes:
```
WT_STAT_RTS_DATA_INCR(session, txn_rts_hs_restore_updates);
```

In `if (dryrun)` doesn't end up sprinkled around the code. I suggest a macro because it also would give us an option to include dry-run versions of those statistics in the future. Which would look something like:

```
#define WT_STAT_RTS_DATA_INCR(session, stat)     \
    do {                                                                                                                    \
        if (!S2C(session)->rts->dryrun)                                                                  \
             WT_STAT_CONN_DATA_INCR(session, stat);                                       \
        else                                                                                                              \
             WT_STAT_CONN_DATA_INCR(session, stat##_dryrun);                      \
    while (0)
```

I might be a bit of, but I think the `##` notation in a macro causes string/name concatenation.

Reply:
Thanks, I really like this idea - done in 31075de.